<!--
%\VignetteEngine{knitr}
%\VignetteIndexEntry{HWxtest}
-->
```{r setup, include=FALSE}
library(knitr)
library(knitcitations)
library(HWxtest)
library(adegenet)
figw <- 8
figh <- 6
set.seed(60823316) 
```
```{r refs, include=FALSE}
#bib <- read.bibtex('~/DropBox/HWxtest/bibHW.txt')
wd <- dirname(getwd())
wd <- paste(wd,"/inst/bibHW.txt",sep="")
bib <- read.bibtex("bibHW.txt")
engels2009 <- ref("10.1534/genetics.109.108977")
levene1949 <- ref("10.1214/aoms/1177730093")
haldane1954 <- ref("10.1007/BF02985085")
louis1987 <- ref("10.2307/2531534")
guo1992 <- ref("10.2307/2532296")
ward2014 <- ref("10.1093/biostatistics/kxt028")
rousset1995 <- bib[["rousset1995"]]
robertson1984 <- bib[["robertson1984"]]
genepop007 <- bib[["genepop007"]]
adegenet <- ref("10.1093/bioinformatics/btn129")
pegas <- ref("10.1093/bioinformatics/btp696")
morin2012 <- bib[["morin2012"]]
```
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 120)
```

<STYLE type="text/css">
  h1,h2,h3,h4,h5 { 
    font-family: palatino, georgia, serif;
    color: Maroon;
  }
    h1, h6{
    text-align: center;
  }
  h1{line-height: 50px}
  body{
    font-size: 0.9em;
    line-height: 23px;
  }
  h3{
  font-weight: normal;
  font-size: 1.8em;
  }
  h6{
        font-size: 0.9em;
        font-weight: normal;
        line-height: 5px;      
   }
   hr{
     border-top-style: solid;
     border-top-width: medium;
   }
  code {
    font-size: 80%;
    line-height: 140%;
    border: 1px solid #ccc;
  }
   @media print{
  hr { 
      visibility: inherit;
      page-break-before: auto;
    }
   }
 </STYLE>


* * *
HWxtest -- Exact Tests for Hardy-Weinberg Proportions With Multiple Alleles
========================================================

###### William R. Engels  --  <wrengels@wisc.edu>  
###### University of Wisconsin, Madison -- Genetics Department
* * *
<center>
* [Main features](#intro)
* [Quick example](#qex)
* [Plot of test statistic](#plot1)
* [Extended example: Bowhead whales](#wex)
* [Inaccuracy in asymptotic tests](#asym)
* [Bibliography](#bib)
</center>

### <a name="intro">Main features</a>
The __HWxtest__ package tests whether a set of genotype counts fits Hardy-Weinberg proportions using the methods described by `r citet(engels2009)`. The package's main features are:
* Performs fast exact test with any number of alleles
* The P value is determined using a choice of test statistics (see below). One option is to test specifically for an excess of homozygotes or heterozygotes.
* Can handle data sets with multiple loci or populations. It accepts data from other population genetics software: *GenePop* `r citep(genepop007)`, *adegenet* `r citep(adegenet)`, *pegas* `r citep(pegas)`
* Can perform either full enumeration test or Monte Carlo, depending on which is optimal for a given data set.
* Includes functions for determining the number of tables for a given set of allele counts

### <a name="qex">Quick example</a>
Suppose you sample 279 individuals from a population and determine their genotypes at a particular 3-allele locus. If the allele names are $a_1, a_2$ and $a_3$, the genotype counts might be:
$$latex
{\bf{a}} = \left[ {\begin{array}{*{20}{c}}
{{a_{11}}}&{}&{}\\
{{a_{21}}}&{{a_{22}}}&{}\\
{{a_{31}}}&{{a_{32}}}&{{a_{33}}}
\end{array}} \right] = \left[ {\begin{array}{*{20}{c}}
{83}&{}&{}\\
{49}&{18}&{}\\
{74}&{34}&{21}
\end{array}} \right]
$$
```{r include=FALSE}
obs <- c(83, 49, 18, 74, 34, 21)
```
That is, we observed `r obs[1]` homozygotes for allele $a_1$, `r obs[2]` heterozygotes of genotype $a_1/a_2$, and so on. The genotype counts are in the lower half of a $3 \times 3$ matrix. (These values come from `r citet(morin2012)`) We'll enter the numbers by rows in an *R* vector:
```{r entera}
obs <- c(83, 49, 18, 74, 34, 21)
```
To test whether these numbers fit the Hardy-Weinberg proportions, call the function `hw.test`
```{r call1}
result <- hw.test(obs)
result
```
The output tells us that there are `r as.integer(result$tableCount)` possible samples of `r sum(obs)` diploids with the same allele counts as the observed data. The P values were computed by looking at each of those possibilities and adding up the total probability of all tables which deviate from HW by at least as much as the observed. 

But why does it report *four* different P values? The reason is that there are several ways one can measure just how deviant a particular outcome is compared to the HW expectation. These measures are explained in more detail below. Briefly, if you had no prior expectation that there might be too many homozygotes, then the first (*LLR*) P value is recommended. However, if you had a prior suspicion that homozygotes would be in excess, such as from inbreeding, population admixture, null alleles, *etc.*, then the third choice (*U*) would be prefereable.

### <a name="plot1">Plot a histogram of the test statistic</a>

The `hw.test` function can also show a plot of how the test statistic is actually distributed, as opposed to its theoretical distribution for large samples. To do this, specify the number of bins to use in the histogram:
```{r plot1, fig.width=figw, fig.height=figh}
hw.test(obs, histobins=T)
```
The red area represents those outcomes found to be more extreme than the observed data by the *LLR* criterion. In other words, the red area represents the P value. The blue curve shows the theoretical $\chi^2$ distribution for comparison. In this case the fit is pretty good.

We can do the same thing for the *U* score by specifying the `statName`:
```{r plot2, fig.width=figw, fig.height=figh}
hw.test(obs, histobins=T, detail=0, statName="U")
```
Note that setting the parameter `detail` to zero suppresses the re-printing of the numbers.

### <a name="wex">Example with real data: bowhead whales</a>
A set of data from P. `r citet(morin2012)` includes genotypes of 280 whales from one population and 49 from another. Each individual was classified at 51 genetic loci. Some of these loci had more than 20 alleles. We can test for Hardy-Weinberg fit for each locus and each population:
```{r testwhales}
data(whales.genind)
wtest <- hw.test(whales.genind)
```
Note that the format of the data is of class `genind` from the package `adegenet` which must be installed to perform the test. The result includes 102 individual tests, so let's not print them all out. Instead, we'll use the function `hwdf` to put the results in the form of a data frame, and look at just the first 10 in the list:
```{r dfwhales}
dfwhales <- hwdf(wtest)
dfwhales[1:10,]
```
The first column shows the population, such as "P1," and locus, such as "BmC5R700_Y9101" of each sample. The second column is the P value from the *LLR* criterion by default. *N* is the number of diploid individuals in the sample and *k* is the number of alleles.

Note that there is a standard error associated with the last two tests, but not the others. This is because those two tests were done by the Monte Carlo method rather than full enumeration. The `hw.test` function decides which method to use based on how many tables would need to be examined in order to do the full enumeration. In these two samples, the number of tables would be `r acount(wtest$P1$Bmys42aK46_R225_K232$genotypes)` and `r as.integer(xcount(wtest$P1$Bmys43Y237_Y377$genotypes))`  respectively, and that would take too long. Instead, `hw.test` performed 100,000 random trials to estimate the P value in each case.

If you need a more precise P value than the one provided by 100,000 trials of the Monte Carlo test, there are two ways to do it. One way is to simply increase the Monte Carlo sample size. Suppose we want to do that for the 7-allele case, P1.Bmys42aK46_R225_K232, but not for the whole data set. We can start by pulling out the genotype counts for that specific test. These genotype counts are contained in the `wtest` object and can be picked out as follows:
```{r getcounts}
counts1 <- wtest$P1$Bmys42aK46_R225_K232$genotypes
counts1
```
Now we can call `hw.test` again, but increase the sample size to one million by changing the parameter `B`:
```{r biggerB}
hw.test(counts1, detail=1, B=1000000)
```
The standard error of the P value is now considerably smaller. For the other sample, P1.Bmys43Y237_Y377, the number of tables is not as huge, and we can obtain an exact P value by increasing the cutoff number. The following tells `hw.test` to perform the full enumeration test instead of the Monte Carlo unless the number of tables exceeds 2 billion:
```{r bigger cutoff}
counts2 <- wtest$P1$Bmys43Y237_Y377$genotypes
hw.test(counts2, detail=1, cutoff=2e8)
```
This time, the P value is reported without a standard error, indicating that an exact test with full enumeration was performed.

The results contained in `wtest` includes P values for all four criteria. So, for example, if we were interested in the *U* test rather than the *LLR* we could have specified that in the data frame call:
```{r Udataframe}
hwdf(wtest, statName="U")[1:10,]
```
When using the *U* test, it is important to pay attention to the sign of *U*. If it is negative, as in the first two samples, the test is specifically for heterozygote excess. If it is positive, then we are testing for homozygote excess. This point will be discussed further below. 

### <a name="asym">Standard asymptotic tests are often inaccurate</a>
Most introductory genetics classes teach that the way to test for Hardy-Weinberg proportions is the standard $\chi^2$ test with $k(k-1)/2$ degrees of freedom. (I am guilty of this myself in teaching Genetics 466 at the University of Wisconsin!) The problem is that with real data and multiple alleles, there are often rare alleles which make the asymptotic test unreliable. 


### <a name="bib">Bibliography</a>
```{r results="asis", echo=FALSE}
bibliography("html")
````
